src/commands/ipmi.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+#==============================================================================
+# Copyright (C) 2017 Stephen F. Norledge and Alces Software Ltd.
+#
+# This file/package is part of Alces Metalware.
+#
+# Alces Metalware is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Affero General Public License
+# as published by the Free Software Foundation, either version 3 of
+# the License, or (at your option) any later version.
+#
+# Alces Metalware is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this package.  If not, see <http://www.gnu.org/licenses/>.
+#
+# For more information on the Alces Metalware, please visit:
+# https://github.com/alces-software/metalware
+#==============================================================================
+
+require 'command_helpers/base_command'
+require 'command_helpers/node_identifier'
+require 'config'
+
+module Metalware
+  module Commands
+    class Ipmi < CommandHelpers::BaseCommand
+      private
+
+      prepend CommandHelpers::NodeIdentifier
+
+      def setup
+        @options = options
+        @args = args
+      end
+
+      def run
+        ipmi(command)
+      end
+
+      def ipmi(cmd)
+        system(cmd.to_s)
+      end
+
+      def node_names
+        @node_names ||= nodes.map(&:name)
+      end
+
+      def command
+        "ipmitool #{render_hosts} #{render_credentials} #{render_command}"
+      end
+
+      def render_command
+        @options.command
+      end
+
+      def render_credentials
+        if @options.group
+          username = alces.domain.config.networks.bmc.bmcuser
+          password = alces.domain.config.networks.bmc.bmcpassword
+        else
+          node = alces.nodes.find_by_name(node_names[0])
+          username = node.config.networks.bmc.bmcuser
+          password = node.config.networks.bmc.bmcpassword
+        end
+        # "-U #{username.to_s} -P #{password.to_s}"
+        "-U admin -P #{password}"
+      end
+
+      def render_hostname
+        node = alces.nodes.find_by_name(node_names[0])
+        node.config.networks.bmc.ip
```

Comment:
Do we want a check that the network is defined?

Reply:
I think this would be a good idea. I think it's fine to just access this part of the config for now as we just have one standard repo and we need to depend on its structure to some extent, but adding a check like this in the `setup` method (before we try to use the config) could be useful as [the BMC network will not always be defined for a node](https://github.com/alces-software/metalware-default/blob/develop/config/domain.yaml#L68):

```ruby
unless object.config.networks.bmc.defined
  raise MetalwareError, "BMC network not defined for #{object.name}"
end
```

Reply:
Have you seen this @vlj91?

Reply:
I've removed `render_hostname`, it's not really required - and put the BMC network defined check into `render_credentials` instead.